src/js/webrequest.js

```diff
@@ -738,6 +738,43 @@ function dispatcher(request, sender, sendResponse) {
     badger.storage.revertUserAction(request.origin);
     sendResponse();
 
+  } else if (request.type == "downloadWhitelist") {
+    chrome.storage.sync.get("disabledSites", function (store) {
+      if (store.hasOwnProperty("disabledSites")) {
+        badger.getSettings().setItem("disabledSites", store.disabledSites);
+        sendResponse();
+      } else {
+        //TODO display a message to the user?
+        console.log("No whitelist in sync storage.");
+      }
+    });
+
+  } else if (request.type == "mergeWhitelist") {
+    chrome.storage.sync.get("disabledSites", function (store) {
+      if (store.hasOwnProperty("disabledSites")) {
+        var settings = badger.getSettings();
+        var whitelist = _.union(settings.getItem("disabledSites"), store.disabledSites);
+        settings.setItem("disabledSites", whitelist);
+        sendResponse();
+      } else {
+        //TODO display a message to the user?
+        console.log("No whitelist in sync storage.");
+      }
+    });
+
+  } else if (request.type == "uploadWhitelist") {
+    var obj = {};
+    obj.disabledSites = badger.getSettings().getItem("disabledSites");
+    chrome.storage.sync.set(obj, function () {
+      if (chrome.runtime.lastError) {
```

Comment:
We don't need to do the `criticalError` stuff for `storage.sync` saving since it will always be kicked off by the user and we will use alerts on the options page to provide feedback.

The `criticalError` system was added to handle reporting issues with updating Badger data to the user (so that they can in turn let us know about the problem): #1848. Badger data can get updated by all sorts of things; usually the updating isn't connected to the user interacting with Privacy Badger at all. In our case though we know the user just clicked a particular button on the options page.